presto-native-sidecar-plugin/src/main/java/com/facebook/presto/sidecar/functionNamespace/NativeFunctionNamespaceManager.java

```diff
@@ -0,0 +1,484 @@
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.MoreCollectors.onlyElement;
+import static java.lang.Long.parseLong;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.HOURS;
+
+public class NativeFunctionNamespaceManager
+        extends AbstractSqlInvokedFunctionNamespaceManager
+{
+    private static final Logger log = Logger.get(NativeFunctionNamespaceManager.class);
+    private final Map<QualifiedObjectName, UserDefinedType> userDefinedTypes = new ConcurrentHashMap<>();
+    private final Map<SqlFunctionHandle, AggregationFunctionImplementation> aggregationImplementationByHandle = new ConcurrentHashMap<>();
+    private final FunctionDefinitionProvider functionDefinitionProvider;
+    private final NodeManager nodeManager;
+    private final Map<SqlFunctionId, SqlInvokedFunction> latestFunctions = new ConcurrentHashMap<>();
+    private final Supplier<Map<SqlFunctionId, SqlInvokedFunction>> memoizedFunctionsSupplier;
+    private final FunctionMetadataManager functionMetadataManager;
+    private final LoadingCache<Signature, SqlFunctionSupplier> specializedFunctionKeyCache;
+    private final List<Type> types;
+
+    @Inject
+    public NativeFunctionNamespaceManager(
+            @ServingCatalog String catalogName,
+            SqlFunctionExecutors sqlFunctionExecutors,
+            SqlInvokedFunctionNamespaceManagerConfig config,
+            FunctionDefinitionProvider functionDefinitionProvider,
+            NodeManager nodeManager,
+            FunctionMetadataManager functionMetadataManager)
+    {
+        super(catalogName, sqlFunctionExecutors, config);
+        this.functionDefinitionProvider = requireNonNull(functionDefinitionProvider, "functionDefinitionProvider is null");
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+        this.memoizedFunctionsSupplier = Suppliers.memoizeWithExpiration(this::bootstrapNamespace,
+                config.getFunctionCacheExpiration().toMillis(), TimeUnit.MILLISECONDS);
+        this.functionMetadataManager = requireNonNull(functionMetadataManager, "functionMetadataManager is null");
+        this.specializedFunctionKeyCache = CacheBuilder.newBuilder()
+                .maximumSize(1000)
+                .expireAfterWrite(1, HOURS)
+                .build(CacheLoader.from(this::doGetSpecializedFunctionKey));
+        this.types = functionMetadataManager.getTypes();
+    }
+
+    private SqlFunctionSupplier doGetSpecializedFunctionKey(Signature signature)
+    {
+        return functionMetadataManager.getSpecializedFunctionKey(signature);
+    }
+
+    private Map<SqlFunctionId, SqlInvokedFunction> bootstrapNamespace()
```

Comment:
Here is how I would suggest the concurrency control to go.

Wherever we mutate `latestFunctions`, synchronize that.  If it's lightweight and in-memory, just synchronize the method.  This applies to `createFunction` and bootstrapNamespace`.